File: Matrix Multiplication/pmp.jl
using LinearAlgebra
using Base.Threads

function parallel_matmul(A::Matrix{Float64}, B::Matrix{Float64})
    m, n = size(A)
    p = size(B, 2)
    C = zeros(m, p)
    
    # Parallelize row computation across available threads
    @threads for i in 1:m
        for j in 1:p
            acc = 0.0
            for k in 1:n
                acc += A[i,k] * B[k,j]
            end
            C[i,j] = acc
        end
    end
    return C
end

function validate(C1::Matrix{Float64}, C2::Matrix{Float64}, tol=1e-6)
    return isapprox(C1, C2, atol=tol)
end

# Matrix dimensions
const N = 1024  # Size of N x N matrices

# Generate random matrices
A = rand(N, N)
B = rand(N, N)

# Warmup (JIT compilation)
parallel_matmul(A, B);

# Parallel multiplication
println("Parallel implementation:")
@time C_parallel = parallel_matmul(A, B)

# Built-in multiplication
println("\nBuilt-in multiplication:")
@time C_builtin = A * B

# Validation
if validate(C_parallel, C_builtin)
    println("\nValidation PASSED")
else
    println("\nValidation FAILED")
end

# Performance comparison
println("\nPerformance ratio (built-in/parallel): $(@elapsed(A*B) / $(@elapsed parallel_matmul(A,B))")

File: Matrix Multiplication/pmp.py
import numpy as np
import multiprocessing as mp
from time import time

def parallel_matrix_mult(A, B, num_workers=None):
    """
    Parallel matrix multiplication using multiprocessing
    A: numpy array of shape (m, n)
    B: numpy array of shape (n, p)
    Returns: numpy array of shape (m, p)
    """
    m, n = A.shape
    n, p = B.shape
    result = np.zeros((m, p))
    
    # Create pool of workers
    with mp.Pool(processes=num_workers) as pool:
        # Calculate chunk size
        chunk_size = m // (num_workers or mp.cpu_count()) or 1
        
        # Create tasks
        tasks = [(A[i:i+chunk_size], B) for i in range(0, m, chunk_size)]
        
        # Process tasks in parallel
        results = pool.starmap(np.dot, tasks)
        
        # Combine results
        result = np.vstack(results)
    
    return result

def validate(C1, C2, atol=1e-6):
    """Validate two matrices using numpy's allclose"""
    return np.allclose(C1, C2, atol=atol)

if __name__ == "__main__":
    # Matrix dimensions
    m, n, p = 512, 512, 512
    
    # Generate random matrices
    A = np.random.randn(m, n)
    B = np.random.randn(n, p)
    
    # Parallel multiplication
    start = time()
    C_parallel = parallel_matrix_mult(A, B, num_workers=mp.cpu_count())
    parallel_time = time() - start
    print(f"Parallel time: {parallel_time:.4f}s")
    
    # NumPy reference implementation
    start = time()
    C_numpy = np.dot(A, B)
    numpy_time = time() - start
    print(f"NumPy time: {numpy_time:.4f}s")
    
    # Validation
    if validate(C_parallel, C_numpy):
        print("Validation PASSED")
    else:
        print("Validation FAILED")
    
    print(f"Parallel/NumPy speed ratio: {parallel_time/numpy_time:.2f}x")
